source/gameengine/Rasterizer/RAS_OpenGLRasterizer/RAS_OpenGLRasterizer.cpp

```diff
@@ -208,6 +210,181 @@ void RAS_OpenGLRasterizer::ScreenPlane::Render()
 	glBindBufferARB(GL_ELEMENT_ARRAY_BUFFER_ARB, 0);
 }
 
+RAS_OpenGLRasterizer::OffScreens::OffScreens()
+	:m_currentIndex(-1),
+	m_width(0),
+	m_height(0),
+	m_samples(0),
+	m_hdr(RAS_HDR_NONE)
+{
+	for (unsigned short i = 0; i < RAS_IRasterizer::RAS_OFFSCREEN_MAX; ++i) {
+		m_offScreens[i] = NULL;
+	}
+}
+
+RAS_OpenGLRasterizer::OffScreens::~OffScreens()
+{
+	for (unsigned short i = 0; i < RAS_IRasterizer::RAS_OFFSCREEN_MAX; ++i) {
+		if (m_offScreens[i]) {
+			GPU_offscreen_free(m_offScreens[i]);
+		}
+	}
+}
+
+GPUOffScreen *RAS_OpenGLRasterizer::OffScreens::GetOffScreen(unsigned short index)
+{
+	if (!m_offScreens[index]) {
+		// The offscreen need to be created now.
+
+		// Check if the FBO index can support samples.
+		const bool sampleofs = index == RAS_OFFSCREEN_RENDER ||
+							   index == RAS_OFFSCREEN_EYE_LEFT0 ||
+							   index == RAS_OFFSCREEN_EYE_RIGHT0;
+
+		// Get FBO mode : render buffer support for multisampled FBOs.
+		int mode = GPU_OFFSCREEN_MODE_NONE;
+		if (sampleofs && (m_samples > 0)) {
+			mode = GPU_OFFSCREEN_RENDERBUFFER_COLOR | GPU_OFFSCREEN_RENDERBUFFER_DEPTH;
+		}
+
+		m_offScreens[index] = GPU_offscreen_create(m_width, m_height, sampleofs ? m_samples : 0, (GPUHDRType)m_hdr, mode, NULL);
```

Comment:
Some hardware doesn't support high MSAA settings combined with HDR settings. Should we make a loop downgrading these settings while there's an error ?